Integration tests for livelog and fix for s3 upload

jonasfj opened the pull request.

jonasfj commented:
Now this also fixes empty artifact upload, issue #375

walac commented on worker/workertest/superseded_test.go:
```diff
@@ -88,42 +35,101 @@ func TestSuperseded(t *testing.T) {
+				// If we haven't made a superseded task yet we do that now
+				if supersededTaskID == "" {
+					supersededTaskID = taskID
+					// Make first other taskId be superseding this taskId
```
I didn't understand this sentence

walac commented on test/nativetest/artifacts_posix_test.go:
```diff
@@ -11,144 +11,144 @@ import (
+						{
+							"name": "public/result.txt",
+							"type": "file",
+							"path": "no-such-file-` + filename + `.txt"
```
Where is the variable filename defined?

jonasfj replied:
Probably higher up...this is just a mechanically refactored test case.. see the commit

walac commented on test/scripttest/logging_test.go:
```diff
@@ -57,47 +67,177 @@ func TestLogging(t *testing.T) {
+				taskID := r.URL.Query().Get("taskId")
+				msg := messages[r.URL.Query().Get("msgSize")]
+				behavior := r.URL.Query().Get("behavior")
+				reply := replies[r.URL.Query().Get("replySize")]
```
Are `msg` and `reply` buffers or buffer sizes?

jonasfj replied:
They are strings... `messages` is a mapping from _string representing size_ to _message as string_.

walac commented on test/scripttest/logging_test.go:
```diff
@@ -57,47 +67,177 @@ func TestLogging(t *testing.T) {
+					}
+				}
+
+				// Signal that eveyrthing went well, and return reply.
```
nit: "everything"

walac commented on worker/workertest/workertest.go:
```diff
@@ -299,31 +299,35 @@ func (c Case) testWithQueue(t *testing.T, q *tcqueue.Queue, l fakequeue.Listener
+	}
+
+	// Create task asynchronously with a limit of 10 concurrent calls to make
+	// that it goes fairly faster, when there is a lot tasks.
```
nit: "a not of"

walac commented on worker/workertest/workertest.go:
```diff
@@ -299,31 +299,35 @@ func (c Case) testWithQueue(t *testing.T, q *tcqueue.Queue, l fakequeue.Listener
+		tdefs[i].Metadata.Owner = "jonasfj@mozilla.com"
+	}
+
+	// Create task asynchronously with a limit of 10 concurrent calls to make
```
nit: "task**s**"

walac commented on worker/workertest/fakequeue/fakequeue.go:
```diff
@@ -529,6 +529,16 @@ func (q *FakeQueue) internalPutArtifact(taskID string, runID int, name string, p
 	// Real S3/azure is slow, let's add a bit of jitter to get some intermittent bugs
 	time.Sleep(time.Duration(rand.Intn(15)) * time.Millisecond) // sleep 0 - 15ms
 
+	// S3 always requires a content-length, if it's not present we have bug
```
nit: "have **a** bug"

jonasfj commented:
@walac, I added a fix for the broken tests in master... due to missing test fixes from the `/` fixes...
A previously commit in this sequence fixes a bug a that hid CI issues in one of the test configurations.

I also rebased and merged a few of the commits... So each commit does one thing.
